File: rest/src/main/java/io/sparkled/viewmodel/playlist/sequence/channel/PlaylistSequenceViewModelConverterImpl.java
package io.sparkled.viewmodel.playlist.sequence.channel;

import com.google.gson.Gson;
import io.sparkled.model.animation.SequenceChannelEffects;
import io.sparkled.model.animation.effect.Effect;
import io.sparkled.model.entity.SequenceChannel;
import io.sparkled.persistence.sequence.SequencePersistenceService;

import javax.inject.Inject;
import java.util.Collections;
import java.util.List;

public class PlaylistSequenceViewModelConverterImpl implements PlaylistSequenceViewModelConverter {

    private SequencePersistenceService sequencePersistenceService;
    private Gson gson = new Gson();

    @Inject
    public PlaylistSequenceViewModelConverterImpl(SequencePersistenceService sequencePersistenceService) {
        this.sequencePersistenceService = sequencePersistenceService;
    }

    @Override
    public PlaylistSequenceViewModel toViewModel(SequenceChannel model) {
        SequenceChannelEffects sequenceChannelEffects = gson.fromJson(model.getChannelJson(), SequenceChannelEffects.class);
        List<Effect> effects = sequenceChannelEffects == null ? Collections.emptyList() : sequenceChannelEffects.getEffects();

        return new PlaylistSequenceViewModel()
                .setUuid(model.getUuid())
                .setSequenceId(model.getSequenceId())
                .setStagePropUuid(model.getStagePropUuid())
                .setName(model.getName())
                .setDisplayOrder(model.getDisplayOrder())
                .setEffects(effects);
    }

    @Override
    public SequenceChannel fromViewModel(PlaylistSequenceViewModel viewModel) {
        SequenceChannel model = sequencePersistenceService.getSequenceChannelByUuid(viewModel.getSequenceId(), viewModel.getUuid())
                .orElseGet(SequenceChannel::new);

        String channelJson = gson.toJson(new SequenceChannelEffects().setEffects(viewModel.getEffects()));
        return model
                .setUuid(viewModel.getUuid())
                .setSequenceId(viewModel.getSequenceId())
                .setStagePropUuid(viewModel.getStagePropUuid())
                .setName(viewModel.getName())
                .setDisplayOrder(viewModel.getDisplayOrder())
                .setChannelJson(channelJson);
    }
}

File: rest/src/main/kotlin/io/sparkled/viewmodel/playlist/sequence/PlaylistSequenceViewModelConverterImpl.kt
package io.sparkled.viewmodel.playlist.sequence

import io.sparkled.model.entity.PlaylistSequence
import io.sparkled.model.util.IdUtils.NO_ID
import io.sparkled.model.util.IdUtils.NO_UUID
import io.sparkled.persistence.playlist.PlaylistPersistenceService
import javax.inject.Inject

class PlaylistSequenceViewModelConverterImpl
@Inject constructor(private val playlistPersistenceService: PlaylistPersistenceService) : PlaylistSequenceViewModelConverter() {

    override fun toViewModel(model: PlaylistSequence): PlaylistSequenceViewModel {
        return PlaylistSequenceViewModel()
            .setUuid(model.getUuid())
            .setSequenceId(model.getSequenceId())
            .setDisplayOrder(model.getDisplayOrder())
    }

    override fun toModel(viewModel: PlaylistSequenceViewModel): PlaylistSequence {
        val model = playlistPersistenceService
            .getPlaylistSequenceByUuid(viewModel.getSequenceId() ?: NO_ID, viewModel.getUuid() ?: NO_UUID)
            ?: PlaylistSequence()

        return model
            .setUuid(viewModel.getUuid())
            .setSequenceId(viewModel.getSequenceId())
            .setDisplayOrder(viewModel.getDisplayOrder())
    }
}
